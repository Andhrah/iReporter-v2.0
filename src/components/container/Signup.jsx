/* eslint-disable react/no-unused-state */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import SignupView from '../presentational/Signup';
import { signup } from '../../actions';
import Navbar from '../presentational/Navbar';

/**
 * @class Signup
 * @extends { React }
 */

class Signup extends Component {
  state = {
    values: {
      firstname: '',
      lastname: '',
      othername: '',
      email: '',
      username: '',
      phoneNumber: '',
      password: '',
      confirmPassword: '',
    },
    errors: {
      signup: '',
      firstname: '',
      lastname: '',
      othername: '',
      email: '',
      username: '',
      phoneNumber: '',
      password: '',
      confirmPassword: '',
    },
    signUpCompleted: false,
  }

  /**
   * @description - This function handles the set the new state
   * when an element state changes in a form field managed by a component,
   * we track it using the onChange attribute
   * @param {object} event - This is the event object
   */

  handleChange = (event) => {
    const { name, value } = event.target;
    const { errors } = this.state;
    this.setState({
      errors: Object.assign({}, errors, {
        [name]: ''
      }),
    });
    this.setState(prevState => ({ values: { ...prevState.values, [name]: value } }));
  }

  /**
   * This function handles the onsubmit event. It triggers validation and sends data to the API
   * @param {object} event - This is the event object
   */

  handleSubmit = (event) => {
    event.preventDefault();
    const { signup: signupUser } = this.props;
    const {
      values: {
        firstname,
        lastname,
        othername,
        username,
        email,
        phoneNumber,
        password,
        confirmPassword
      }
    } = this.state;

    const userData = {
      firstname,
      lastname,
      othername,
      username,
      email,
      phoneNumber,
      password,
      confirmPassword,
    };

    const hasError = this.handleValidation(userData);
    if (!hasError) {
      signupUser(userData);
      const { history } = this.props;
      this.setState({ signUpCompleted: true, });
      history.push('/');
    }
  }

  /**
   * This function validates data from the form
   * @param {object} data - This is data gotten from the form
   */
  handleValidation = (userData) => {
    const errors = {};
    if (!userData.firstname || userData.firstname.length < 3) {
      errors.firstname = 'First Name should not be less than 3 letters';
    }
    if (!userData.lastname || userData.lastname.length < 3) {
      errors.lastname = 'Last Name should not be less than 3 letters';
    }
    if (!userData.username || userData.username.length < 3) {
      errors.username = 'Username should not be less than 3 digits';
    }
    if (!userData.phoneNumber) {
      errors.phoneNumber = 'Please enter a valid phone number';
    }
    if (!userData.email) {
      errors.email = 'Please enter a valid email';
    }
    if (userData.password !== userData.confirmPassword) {
      errors.password = 'password mismatch';
      errors.confirmPassword = 'password mismatch';
    }
    if (!userData.password || userData.password.length < 6) {
      errors.password = 'Please enter a valid password, password must not be less than 6 characters';
    }
    this.setState({ errors });
    return Object.keys(errors).some(field => Boolean(field));
  }

  render() {
    const { values, errors } = this.state;
    return (
      <>
        <Navbar />
        <SignupView
        handleChange={this.handleChange}
        handleSubmit={this.handleSubmit}
        values={values}
        errors={errors}
      />
      </>
    );
  }
}

Signup.propTypes = {
  signup: PropTypes.func,
  history: PropTypes.object.isRequired,
};

Signup.defaultProps = {
  signup: () => {},
};


export default connect(null, { signup })(withRouter(Signup));
