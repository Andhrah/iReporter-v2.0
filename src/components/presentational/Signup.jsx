import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';


const Signup = ({
  handleChange, handleSubmit, values, errors
}) => {
  const {
    firstname,
    lastname,
    othername,
    email,
    username,
    phoneNumber,
    password,
    confirmPassword,
  } = values;
  const {
    // Signup: signupError,
    firstname: firstnameError,
    lastname: lastnameError,
    othername: othernameError,
    email: emailError,
    username: usernameError,
    phoneNumber: phoneNumberError,
    password: passwordError,
    confirmPassword: confirmPasswordError,
  } = errors;
  return (
    <div className="wrapper">
      <div className="form-container" id="signup-wrapper">
        <h1>Sign Up</h1>
        <form onSubmit={handleSubmit}>
          <div className="form-container-grid">
            {/* <div className="error" id="signup-error">{signupError}</div> */}
            <div className="input-container">
              <label htmlFor="firstname">
                <span>*</span>
              First Name
              </label>
              <input type="text" value={firstname} name="firstname" onChange={handleChange} className="text-input" id="firstname" />
              <div className="error" id="firstname-error">{firstnameError}</div>
            </div>
            <div className="input-container">
              <label htmlFor="lastname">
                <span>*</span>
              Last Name
              </label>
              <input type="text" value={lastname} name="lastname" onChange={handleChange} className="text-input" id="lastname" />
              <div className="error" id="lastname-error">{lastnameError}</div>
            </div>
            <div className="input-container">
              <label htmlFor="othername">Other Name</label>
              <input type="text" value={othername} name="othername" onChange={handleChange} className="text-input" id="othername" />
              <div className="error" id="othername-error">{othernameError}</div>
            </div>
            <div className="input-container">
              <label htmlFor="email">
                <span>*</span>
              Email
              </label>
              <input type="email" value={email} name="email" onChange={handleChange} className="text-input" id="email" />
              <div className="error" id="email-error">{emailError}</div>
            </div>
            <div className="input-container">
              <label htmlFor="phoneNumber">
                <span>*</span>
              Phone Number
              </label>
              <input type="tel" title="Invalid phone number, please follow this pattern *********** e.g  08066677756" pattern="[0-9]{11}" value={phoneNumber} name="phoneNumber" onChange={handleChange} className="text-input" id="phone-number" />
              <div className="error" id="phone-number-error">{phoneNumberError}</div>
            </div>
            <div className="input-container">
              <label htmlFor="username">
                <span>*</span>
              Username
              </label>
              <input type="text" value={username} name="username" onChange={handleChange} className="text-input" id="username" />
              <div className="error" id="username-error">{usernameError}</div>
            </div>
            <div className="input-container">
              <label htmlFor="password">
                <span>*</span>
              Password
              </label>
              <input type="password" value={password} name="password" onChange={handleChange} className="text-input" id="password" />
              <div className="error" id="password-error">{passwordError}</div>
            </div>
            <div className="input-container">
              <label htmlFor="confirmPassword">
                <span>*</span>
              Confirm Password
              </label>
              <input type="password" value={confirmPassword} name="confirmPassword" onChange={handleChange} className="text-input" id="confirm-password" />
              <div className="error" id="confirm-password-error">{confirmPasswordError}</div>
            </div>
          </div>
          <button type="submit" className="form-primary-btn" id="signup">
          Sign Up
          </button>
        </form>
        <div className="second-btn">
          <h4 className="form-h4">Already have an account?</h4>
        </div>
        <p className="repeating-linear">
          <Link to="/login" className="form-secondary-btn">
          Sign In
          </Link>
        </p>
      </div>
    </div>
  );
};

Signup.propTypes = {
  handleChange: PropTypes.func,
  handleSubmit: PropTypes.func,
  values: PropTypes.object,
  errors: PropTypes.object,
};

Signup.defaultProps = {
  handleChange: () => {},
  handleSubmit: () => {},
  values: {},
  errors: {},
};

export default Signup;
