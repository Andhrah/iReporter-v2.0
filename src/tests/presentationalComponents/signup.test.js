import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import Signup from '../../components/presentational/Signup';

const props = {
  handleChange: () => {},
  handleSubmit: () => {},
  values: {
    economy: 'economy',
    badroad: 'badroad',
    badHousing: 'badHousing',
    noElectricity: 'noElectricity',
    poorElectricity: 'poorElectricity',
    refuseDamping: 'refuseDamping',
    collapsedBridge: 'collapsedBridge',
    otherReasons: 'nothing',
    otherReasonsSpecify: 'nothing',
    interventionDate: '2019-05=04',
    describe: 'describe',
    interventionImages: 'interventionImages.png',
  },
  errors: {
    emailError: 'invalid email',
    passwordError: 'inavlid password',
  }
};

describe('Signup  test suite', () => {
  it('renders correctly', () => {
    const wrapper = mount(
      <MemoryRouter>
        <Signup {...props} />
      </MemoryRouter>
    );
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });
});
