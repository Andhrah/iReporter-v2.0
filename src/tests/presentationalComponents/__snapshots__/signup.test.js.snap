// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup  test suite renders correctly 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "b40vzh",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "b40vzh",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Signup
      errors={
        Object {
          "emailError": "invalid email",
          "passwordError": "inavlid password",
        }
      }
      handleChange={[Function]}
      handleSubmit={[Function]}
      values={
        Object {
          "badHousing": "badHousing",
          "badroad": "badroad",
          "collapsedBridge": "collapsedBridge",
          "describe": "describe",
          "economy": "economy",
          "interventionDate": "2019-05=04",
          "interventionImages": "interventionImages.png",
          "noElectricity": "noElectricity",
          "otherReasons": "nothing",
          "otherReasonsSpecify": "nothing",
          "poorElectricity": "poorElectricity",
          "refuseDamping": "refuseDamping",
        }
      }
    >
      <div
        className="wrapper"
      >
        <div
          className="form-container"
          id="signup-wrapper"
        >
          <h1>
            Sign Up
          </h1>
          <form
            onSubmit={[Function]}
          >
            <div
              className="form-container-grid"
            >
              <div
                className="input-container"
              >
                <label
                  htmlFor="firstname"
                >
                  <span>
                    *
                  </span>
                  First Name
                </label>
                <input
                  className="text-input"
                  id="firstname"
                  name="firstname"
                  onChange={[Function]}
                  type="text"
                />
                <div
                  className="error"
                  id="firstname-error"
                />
              </div>
              <div
                className="input-container"
              >
                <label
                  htmlFor="lastname"
                >
                  <span>
                    *
                  </span>
                  Last Name
                </label>
                <input
                  className="text-input"
                  id="lastname"
                  name="lastname"
                  onChange={[Function]}
                  type="text"
                />
                <div
                  className="error"
                  id="lastname-error"
                />
              </div>
              <div
                className="input-container"
              >
                <label
                  htmlFor="othername"
                >
                  Other Name
                </label>
                <input
                  className="text-input"
                  id="othername"
                  name="othername"
                  onChange={[Function]}
                  type="text"
                />
                <div
                  className="error"
                  id="othername-error"
                />
              </div>
              <div
                className="input-container"
              >
                <label
                  htmlFor="email"
                >
                  <span>
                    *
                  </span>
                  Email
                </label>
                <input
                  className="text-input"
                  id="email"
                  name="email"
                  onChange={[Function]}
                  type="email"
                />
                <div
                  className="error"
                  id="email-error"
                />
              </div>
              <div
                className="input-container"
              >
                <label
                  htmlFor="phoneNumber"
                >
                  <span>
                    *
                  </span>
                  Phone Number
                </label>
                <input
                  className="text-input"
                  id="phone-number"
                  name="phoneNumber"
                  onChange={[Function]}
                  pattern="[0-9]{11}"
                  title="Invalid phone number, please follow this pattern *********** e.g  08066677756"
                  type="tel"
                />
                <div
                  className="error"
                  id="phone-number-error"
                />
              </div>
              <div
                className="input-container"
              >
                <label
                  htmlFor="username"
                >
                  <span>
                    *
                  </span>
                  Username
                </label>
                <input
                  className="text-input"
                  id="username"
                  name="username"
                  onChange={[Function]}
                  type="text"
                />
                <div
                  className="error"
                  id="username-error"
                />
              </div>
              <div
                className="input-container"
              >
                <label
                  htmlFor="password"
                >
                  <span>
                    *
                  </span>
                  Password
                </label>
                <input
                  className="text-input"
                  id="password"
                  name="password"
                  onChange={[Function]}
                  type="password"
                />
                <div
                  className="error"
                  id="password-error"
                />
              </div>
              <div
                className="input-container"
              >
                <label
                  htmlFor="confirmPassword"
                >
                  <span>
                    *
                  </span>
                  Confirm Password
                </label>
                <input
                  className="text-input"
                  id="confirm-password"
                  name="confirmPassword"
                  onChange={[Function]}
                  type="password"
                />
                <div
                  className="error"
                  id="confirm-password-error"
                />
              </div>
            </div>
            <button
              className="form-primary-btn"
              id="signup"
              type="submit"
            >
              Sign Up
            </button>
          </form>
          <div
            className="second-btn"
          >
            <h4
              className="form-h4"
            >
              Already have an account?
            </h4>
          </div>
          <p
            className="repeating-linear"
          >
            <Link
              className="form-secondary-btn"
              to="/login"
            >
              <a
                className="form-secondary-btn"
                href="/login"
                onClick={[Function]}
              >
                Sign In
              </a>
            </Link>
          </p>
        </div>
      </div>
    </Signup>
  </Router>
</MemoryRouter>
`;
